Add config flow to ecobee

marthoc opened the pull request:
## Breaking Change:

Ecobee will now be set up via config flow. Existing users will have their config imported from ecobee.conf via an import flow so it shouldn't break their experience. Users configuring via configuration.yaml will have their api key and options imported into the flow but will still need to finish authorization via the flow (instead of the configurator component as previously).

The configuration parameter `hold_temp` has been removed, as it was not being used in the climate platform and had no effect on whether the temperature was held indefinitely or not. Users will need to remove the parameter `hold_temp` from configuration.yaml.

Ecobee-specific services will now be registered under the `ecobee` domain rather than the `climate` domain, and service names will not include the prefix "ecobee_" (e.g. the service "climate.ecobee_resume_program" will become "ecobee.resume_program").

## Description:

Adds a config flow to the ecobee component. This PR is Phase 1 of my planned updates to the ecobee component to make it a first-class integration - in this PR I've tried to make only the minimal changes that were necessary to support loading ecobee via a config entry to aid reviewability.
 
Phase 2 will follow the merging of this PR and add support for the entity and device registries. Phase 3 will make everything async. 

Upstream library has been updated with needed changes for the config flow.

Tests have been added for the config flow.

**Related issue (if applicable):** N/A

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** https://github.com/home-assistant/home-assistant.io/pull/10428

## Example entry for `configuration.yaml` (if applicable): N/A

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

marthoc committed:
Add basic config flow

marthoc committed:
Fix json files

marthoc committed:
Update __init__.py

marthoc committed:
Fix json errors

marthoc committed:
Move constants to const.py

marthoc committed:
Add ecobee to generated config flows

marthoc committed:
Update config_flow for updated API

marthoc committed:
Update manifest to include new dependencies

Bump pyecobee, add aiofiles.

marthoc committed:
Update constants for ecobee

marthoc committed:
Modify ecobee setup to use config flow

marthoc committed:
Bump dependency

marthoc committed:
Update binary_sensor to use config_entry

marthoc committed:
Update sensor to use config_entry

marthoc committed:
Update __init__.py

marthoc committed:
Update weather to use config_entry

marthoc committed:
Update notify.py

marthoc committed:
Update ecobee constants

marthoc committed:
Update climate to use config_entry

marthoc committed:
Avoid a breaking change on ecobee services

marthoc committed:
Store api key from old config entry

marthoc committed:
Allow unloading of config entry

marthoc committed:
Show user a form before import

marthoc committed:
Refine import flow

MartinHjelmare commented on homeassistant/components/ecobee/__init__.py:
```diff
@@ -1,123 +1,167 @@
 
-    setup_ecobee(hass, NETWORK.ecobee, config)
+    for platform in ECOBEE_PLATFORMS:
+        await hass.config_entries.async_forward_entry_unload(config_entry, platform)
```
Use `asyncio.gather` to check that all platforms were unloaded concurrently. Something like this:
```py
tasks = []
for platform in ECOBEE_PLATFORMS:
    tasks.append(
        hass.config_entries.async_forward_entry_unload(config_entry, platform)
    )

return all(await asyncio.gather(*tasks))
```

MartinHjelmare commented on homeassistant/components/ecobee/binary_sensor.py:
```diff
@@ -1,37 +1,42 @@
+from .const import DOMAIN
 
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
```
Please convert this to the coroutine version `async_setup_platform`. It's a bit faster.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,173 @@
+"""Config flow to configure ecobee."""
+import voluptuous as vol
```
Please sort imports and put 3rd party import group between standard library and homeassistant import groups.

marthoc replied:
Is this the case for all imports from pyecobee? Should all pyecobee imports be imported at the top of the various files? I know it used to be the case that 3rd party imports needed to be inside the functions that used them.

MartinHjelmare replied:
Yes all 3rd party imports including pyecobee should be put at the top. After we introduced manifest.json we can import everything at the module level.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,173 @@
+)
+
+
+@config_entries.HANDLERS.register(DOMAIN)
```
Please remove the decorator. It's not needed when we pass the domain as keyword argument.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,173 @@
+        """Handle a flow initialized by the user."""
+        return await self.async_step_init(user_input)
+
+    async def async_step_init(self, user_input=None):
```
The init step is legacy. Please replace it with the user step.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,173 @@
+            else None
+        )
+
+        if self._async_current_entries():
```
Move this first in the method body.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,173 @@
+            if await self.hass.async_add_executor_job(self._ecobee.request_pin):
+                """We have a PIN; move to the next step of the flow."""
+                return await self.async_step_authorize()
+            else:
```
We can remove `else:` and outdent below since we return above.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,173 @@
+                    CONF_REFRESH_TOKEN: self._ecobee.refresh_token,
+                }
+                return self.async_create_entry(title=DOMAIN, data=config)
+            else:
```
As above.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,173 @@
+        config_file = self.hass.config.path(ECOBEE_CONFIG_FILENAME)
+
+        async with aiofiles.open(config_file, "r") as f:
+            legacy_config = json.loads(await f.read())
```
We have `load_json` util to load json files. Please use that instead. Schedule it on the executor thread pool.

https://github.com/home-assistant/home-assistant/blob/1d3f2d20d2eb2d12d02cde93d7d79e66f8157a1c/homeassistant/util/json.py#L22

MartinHjelmare commented on homeassistant/components/ecobee/manifest.json:
```diff
@@ -1,10 +1,9 @@
+    "config_flow": true,
+    "documentation": "https://www.home-assistant.io/components/ecobee",
+    "dependencies": [],
+    "requirements": ["python-ecobee-api==0.1.2", "aiofiles"],
```
We can remove aiofiles.

MartinHjelmare commented on homeassistant/components/ecobee/sensor.py:
```diff
@@ -59,7 +63,14 @@ def device_class(self):
+        from pyecobee.const import ECOBEE_STATE_CALIBRATING, ECOBEE_STATE_UNKNOWN
+
+        if self._state in [ECOBEE_STATE_CALIBRATING, ECOBEE_STATE_UNKNOWN]:
+            return STATE_UNKNOWN
```
Use `None` to represent unknown state.

MartinHjelmare commented on homeassistant/components/ecobee/strings.json:
```diff
@@ -0,0 +1,39 @@
+                "title": "Authorize app on ecobee.com",
+                "description": "Please authorize this app at https://www.ecobee.com/consumerportal/index.html with pin code:\n\n{pin}\n\nThen, press Submit."
+            },
+            "import": {
```
We usually don't add strings for the import step. Where is it used?

MartinHjelmare commented on homeassistant/components/ecobee/strings.json:
```diff
@@ -0,0 +1,39 @@
+        "abort": {
+            "credentials_not_found": "No valid credentials found in ecobee.conf; please re-start configuration.",
+            "one_instance_only": "This integration currently supports only one ecobee instance.",
+            "refresh_token_expired": "Your refresh token has expired; please re-start configuration."
```
Where is this string used?

MartinHjelmare reviewed (changes requested):
The config flow module requires 100 % test coverage. Please see eg hue config flow tests for examples and make sure that the config flow module isn't excluded from coverage calculation.

marthoc committed:
Update strings.json to remove import step

Not needed.

marthoc committed:
Move third party imports to top of module

marthoc committed:
Remove periods from end of log messages

marthoc commented on homeassistant/components/ecobee/__init__.py:
```diff
@@ -1,123 +1,167 @@
+    But, an existing configuration.yaml config and ecobee.conf
+    will trigger an import flow if a config entry doesn't
+    already exist to help users migrating from the old ecobee
+    component.
```
OK. I think we should always trigger the import flow then if there is an entry in configuration.yaml and a config entry doesn't already exist. The import step will then differentiate between cases where ecobee.conf exists (attempt to automatically import) and when it doesn't (delegate to the user step to continue setup).

marthoc committed:
Make configuration.yaml config optional

marthoc committed:
Remove unused strings

marthoc committed:
Reorganize config flow

marthoc committed:
Remove unneeded requirement

marthoc committed:
No need to store API key

marthoc committed:
Update async_unload_entry

marthoc committed:
Clean up if/else statements

MartinHjelmare commented on homeassistant/components/ecobee/__init__.py:
```diff
@@ -1,123 +1,162 @@
             {
                 vol.Optional(CONF_API_KEY): cv.string,
-                vol.Optional(CONF_HOLD_TEMP, default=False): cv.boolean,
+                vol.Optional(CONF_HOLD_TEMP, DEFAULT_HOLD_TEMP): cv.boolean,
```
`default=DEFAULT_HOLD_TEMP`

MartinHjelmare commented on homeassistant/components/ecobee/__init__.py:
```diff
@@ -1,123 +1,162 @@
+    migrating from the old ecobee component. Otherwise, the user will have to
+    continue setting up the integration via the config flow.
+    """
+    hass.data[DATA_ECOBEE_CONFIG] = config[DOMAIN]
```
If the user triggers the component setup from the GUI there will be no config section for this domain.
```py
hass.data[DATA_ECOBEE_CONFIG] = config.get(DOMAIN, {})
```

MartinHjelmare commented on homeassistant/components/ecobee/binary_sensor.py:
```diff
@@ -1,37 +1,42 @@
+    data = hass.data[DOMAIN]
     dev = list()
     for index in range(len(data.ecobee.thermostats)):
         for sensor in data.ecobee.get_remote_sensors(index):
```
Is this returning cached results or doing I/O?

marthoc replied:
Returning cached results. The update() method of the data object does I/O and the response is cached  in data.ecobee.thermostats. get_remote_sensors() parses the data.ecobee.thermostats property but does not do I/O.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,160 @@
+        errors = {}
+        stored_api_key = (
+            self.hass.data[DATA_ECOBEE_CONFIG].get(CONF_API_KEY)
+            if DATA_ECOBEE_CONFIG in self.hass.data
```
We don't need this check now.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,160 @@
+                        CONF_REFRESH_TOKEN: ecobee.refresh_token,
+                    },
+                )
+        except Exception:  # pylint: disable=broad-except
```
We should catch one or more specific errors. What error are we expecting?

Please wrap as little as possible in each try... except. Move the except statement up under loading json if that's where the first error can happen.

marthoc replied:
The only real expected exception is a KeyError if legacy_config is {} (which could happen because ecobee.conf doesn't exist or is malformed. But I noticed that load_json can raise other exceptions too. Here I used the broad except because it doesn't really matter for the purposes of the config flow import step _why_ the import failed, only that it did, because then we have to delegate to the user step. And testing if ecobee.refresh_tokens is true should only happen if the ecobee object was created (i.e. no exception had been thrown up to that point).

MartinHjelmare replied:
Please catch `KeyError` and `HomeAssistantError` and move the except up under `load_json`.

marthoc replied:
The KeyError would be raised by the following if either ECOBEE_API_KEY or ECOBEE_REFRESH_TOKEN aren't present in the legacy config.
```
ecobee = Ecobee(
                config={
                    ECOBEE_API_KEY: legacy_config[ECOBEE_API_KEY],
                    ECOBEE_REFRESH_TOKEN: legacy_config[ECOBEE_REFRESH_TOKEN],
                }
            )
```
If I move the except under that clause, then the next IF statement will raise an exception if the exception was raised previously since the ecobee object won't exist. I think I should leave the code in place as is but I can except on only KeyError and HomeAssistantError  rather than broad except. Does that work?

MartinHjelmare commented on homeassistant/components/ecobee/__init__.py:
```diff
@@ -1,123 +1,162 @@
+async def async_setup_entry(hass, entry):
+    """Set up ecobee via a config entry."""
+    if DATA_ECOBEE_CONFIG not in hass.data:
+        hass.data[DATA_ECOBEE_CONFIG] = {}
```
We can remove this. It's done above already.

MartinHjelmare commented on homeassistant/components/ecobee/__init__.py:
```diff
@@ -1,123 +1,162 @@
+    """
+    hass.data[DATA_ECOBEE_CONFIG] = config[DOMAIN]
+
+    if not hass.config_entries.async_entries(DOMAIN):
```
Add a check if the config isn't empty. If it's empty, we shouldn't import, since then we're triggered by the user in GUI.

MartinHjelmare commented:
Please run `script/gen_requirements_all.py`.

marthoc committed:
Update requirements_all.txt

marthoc committed:
Fix config schema

marthoc committed:
Update __init__.py

marthoc committed:
Remove check for DATA_ECOBEE_CONFIG

marthoc committed:
Remove redundant check

marthoc committed:
Add check for DATA_ECOBEE_CONFIG

marthoc committed:
Change setup_platform to async

marthoc committed:
Fix state unknown and imports

marthoc commented on homeassistant/components/ecobee/__init__.py:
```diff
@@ -1,123 +1,167 @@
+
+        _LOGGER.debug("Refreshing ecobee tokens and updating config entry.")
+        if self.ecobee.refresh_tokens():
+            self._hass.config_entries.async_update_entry(
```
Is the fix to wrap async_update_entry with asyncio.run_coroutine_threadsafe?

MartinHjelmare replied:
I would probably make all methods in `EcobeeData` coroutines and schedule the sync calls on the executor in those methods.

Since it's only two sync calls to update the api, and the results are cached, we can easily convert the `update` method in the entities to `async_update`.

MartinHjelmare replied:
How do the actuator method in the entities work? Do they do I/O?

marthoc replied:
The entities all call EcobeeData.update() in their own update methods, so those would need to be converted to async_update if EcobeeData.update was changed to be a coroutine. The climate platform also has several calls in various methods to other aspects of pyecobee that are also doing I/O in order to change thermostat settings.

MartinHjelmare replied:
Ok. My suggestion should work then. We just convert the entity update methods to async_update coroutines and convert the update and refresh methods in `EcobeeData` to coroutines. Schedule the api calls in those two methods on the executor.

Actuator methods are left as is.

marthoc committed:
Change init step to user

marthoc committed:
Have import step raise specific exceptions

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+                        CONF_REFRESH_TOKEN: ecobee.refresh_token,
+                    },
+                )
+        except (HomeAssistantError, KeyError):
```
Please move the except up closer to the potential errors and return to the user step inside that block. Put it before we call ecobee.refresh_tokens. Then we can outdent below.

marthoc replied:
OK I've given it another try.

marthoc committed:
Rearrange try/except block in import flow

marthoc committed:
Convert update() and refresh() to coroutines

...and update platforms to use async_update coroutine.

marthoc committed:
Finish converting init to async

marthoc committed:
Preliminary tests

marthoc committed:
Test full implementation

marthoc committed:
Update test_config_flow.py

MartinHjelmare commented:
Please solve the merge conflict.

MartinHjelmare commented on tests/components/ecobee/test_config_flow.py:
```diff
@@ -0,0 +1,65 @@
+    flow = config_flow.EcobeeFlowHandler()
+    flow.hass = hass
+
+    with patch.object(hass.config_entries, "async_entries", return_value=[{}]):
```
Add a `MockConfigEntry` to hass instead.

https://github.com/home-assistant/home-assistant/blob/20e61fb41b9eb6d7b79862ee39aa202dc344dc7f/tests/common.py#L654

MartinHjelmare commented on tests/components/ecobee/test_config_flow.py:
```diff
@@ -0,0 +1,65 @@
+    mock_ecobee = Mock()
+    mock_ecobee.request_pin.return_value = True
+    mock_ecobee.return_value.pin = "test-pin"
+    config_flow.Ecobee = mock_ecobee
```
Patch the `Ecobee` class in the config flow module instead, with this mock as return value.

MartinHjelmare commented on tests/components/ecobee/test_config_flow.py:
```diff
@@ -0,0 +1,65 @@
+
+    mock_ecobee = Mock()
+    mock_ecobee.request_pin.return_value = False
+    config_flow.Ecobee = mock_ecobee
```
As above.

frenck commented:
Thank you for your contribution thus far! 🎖 Since this is a significant contribution, we would appreciate you'd added yourself to the list of code owners for this integration. ❤️ 

For more information about "code owners", see: [Architecture Decision Record 0008: Code owners](https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md).

marthoc committed:
Update test_config_flow.py

marthoc committed:
Add self to codeowners

marthoc commented:
@MartinHjelmare I changed the instances of Mock() in the tests to patches - if you have a minute, would you mind taking a look at the last test? I can't get the test_pin_request_fails to pass and I'm not sure why - I'm mocking the result of Ecobee.request_pin() to False, which should cause async_step_user to show a form with an error, but it doesn't do that and just proceeds to async_step_authorize. That method gets wrapped in hass.async_add_executor_job as follows:

```
 self._ecobee = Ecobee(config={ECOBEE_API_KEY: user_input[CONF_API_KEY]})

            if await self.hass.async_add_executor_job(self._ecobee.request_pin):
                """We have a PIN; move to the next step of the flow."""
                return await self.async_step_authorize()
            errors["base"] = "pin_request_failed"
```
I'm new to unites so I'm sure I'm missing something obvious but I'm not sure what...

MartinHjelmare commented on tests/components/ecobee/test_config_flow.py:
```diff
@@ -0,0 +1,63 @@
+    flow.hass.data[DATA_ECOBEE_CONFIG] = {}
+
+    with patch("homeassistant.components.ecobee.config_flow.Ecobee") as mock_ecobee:
+        mock_ecobee.request_pin.return_value = True
```
Is it on the return value of `mock_ecobee` we should set the `request_pin` return value?

marthoc replied:
That’s it! Pushed fixes. Still need to test import flow.

marthoc committed:
Update test_config_flow.py

marthoc committed:
Use MockConfigEntry

marthoc committed:
Update test_config_flow.py

MartinHjelmare commented:
Please solve the merge conflict.

marthoc committed:
Update CODEOWNERS

MartinHjelmare commented on homeassistant/components/ecobee/climate.py:
```diff
@@ -163,15 +167,15 @@ def resume_program_set_service(service):
-    hass.services.register(
-        DOMAIN,
+    hass.services.async_register(
+        CLIMATE_DOMAIN,
```
We shouldn't register custom services under the platform domain but under the integration domain, ie ecobee.

marthoc replied:
Sounds good to me.

marthoc committed:
pylint fixes

marthoc committed:
Register services under ecobee domain

Breaking change!

marthoc committed:
Pylint fixes

marthoc committed:
Pylint fixes

marthoc committed:
Pylint fixes

MartinHjelmare commented on homeassistant/components/ecobee/__init__.py:
```diff
@@ -110,7 +110,9 @@ def __init__(self, hass, entry, api_key, refresh_token):
         except ExpiredTokenError:
-            _LOGGER.warn("ecobee update failed; attempting to refresh expired tokens")
+            _LOGGER.warning(
+                "ecobee update failed; attempting to refresh expired tokens"
```
Please start logging messages with capital letter.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,156 @@
+"""Config flow to configure ecobee."""
+from copy import copy
+import voluptuous as vol
```
Please add a blank line between standard library and 3rd party import groups.

MartinHjelmare commented:
Please move the service descriptions to a `services.yaml` file in the ecobee integration package:
https://github.com/home-assistant/home-assistant/blob/5a4a3e17cc820d28520c47d478b19182527547a2/homeassistant/components/climate/services.yaml#L75-L93

marthoc committed:
Move service strings to ecobee domain

marthoc committed:
Fix log message capitalization

marthoc committed:
Fix import formatting

MartinHjelmare reviewed (approved):
Looks great!

MartinHjelmare reviewed (changes requested):
Sorry, one more thing. Please update `.coveragerc` and don't exclude the config flow module.

MartinHjelmare commented:
Also please update the breaking change section to mention the change of service names.

marthoc committed:
Update .coveragerc

marthoc commented:
> Sorry, one more thing. Please update `.coveragerc` and don't exclude the config flow module.

I've updated it - will just need to make sure that coverage on the config_flow is 100%.

MartinHjelmare commented on .coveragerc:
```diff
@@ -156,7 +156,10 @@ omit =
     homeassistant/components/ebusd/*
     homeassistant/components/ecoal_boiler/*
-    homeassistant/components/ecobee/*
+    homeassistant/components/ecobee/binary_sensor.py
```
Please exclude the init module too.

marthoc replied:
Just checked codecov and realized I forgot to add tests for the options flow... will add those tonight.

marthoc committed:
Add __init__ to coveragerc

marthoc committed:
Add option flow test

marthoc committed:
Update .coveragerc

MartinHjelmare commented:
Currently we don't listen for updated entry options. We should probably do that to have a change in options change the behavior of the integration without restarting.

https://developers.home-assistant.io/docs/en/config_entries_options_flow_handler.html#signal-updates

We could do this in a later PR, but then we should note this limitation in the PR description.

marthoc commented:
> Currently we don't listen for updated entry options. We should probably do that to have a change in options change the behavior of the integration without restarting.

I think this should be easy enough to achieve - I'm going to push an update for you to take a look at in a few minutes...

marthoc committed:
Act on updated options

marthoc commented:
@MartinHjelmare OK I pushed a new commit that adds an async listener method to the Thermostat entity class, and that listener gets registered to the config_entry as part of async_setup_entry in the climate platform - seems fairly straightforward. When the entry is updated and the listener is called, the hold_temp option in the entity will be updated. Can you take a look and let me know what you think?

MartinHjelmare commented on homeassistant/components/ecobee/climate.py:
```diff
@@ -114,20 +110,27 @@
+
+    for index in range(len(data.ecobee.thermostats)):
+        dev = Thermostat(data, index, hold_temp)
+        config_entry.add_update_listener(dev.option_update_listener)
```
Please subscribe the listener in entity coroutine method `async_added_to_hass`.

We should unsubscribe the listener in `async_will_remove_from_hass`. `add_update_listener` returns a function to unsubscribe the listener.

marthoc replied:
OK so I was just thinking about that - if I need to subscribe in async_added_to_hass, that means I need to pass the config entry into the instance... if I'm doing that, can't we just save some trouble and do something like:
```python
@property
def hold_temp(self):
    return self._config_entry.options[CONF_HOLD_TEMP]
```
In which case, the hold_temp property will always return the updated option value.

MartinHjelmare replied:
Yes, that should work.

MartinHjelmare commented on homeassistant/components/ecobee/climate.py:
```diff
@@ -206,16 +209,20 @@ def __init__(self, data, thermostat_index, hold_temp):
+    async def option_update_listener(self, hass, entry):
+        """Update relevant options in this entity."""
+        _LOGGER.debug("Updating options in ecobee thermostat {}".format(self.name))
+        self.hold_temp = entry.options[CONF_HOLD_TEMP]
```
Where are we using `self.hold_temp`?

marthoc replied:
You know I was just looking at that very thing and it turns out... it's not being used??

Looking at the previous code, there is no mention of self.hold_temp anywhere in the entity class, and that parameter (hold_temp) is only passed to the climate platform. I suspect it should find it's way into the entity method `hold_preference` - but it's even unclear how given the comments on that method.

My guess (I haven't looked at the commit history) is that hold_temp was once a part of hold_preference, but then hold_preference was changed and hold_temp just hung around.

This makes things easier - drop that config parameter, drop the options flow.

marthoc replied:
I just pushed a batch of commits that removes all references to hold_temp and takes out the options flow code (since there are no other options to set for the ecobee integration at this time).

marthoc committed:
Revert "Act on updated options"

This reverts commit 56b0a859f2e3e80b6f4c77a8f784a2b29ee2cce9.

marthoc committed:
Remove hold_temp from climate

marthoc committed:
Remove hold_temp and options from init

marthoc committed:
Remove options handler from config flow

marthoc committed:
Remove options strings

marthoc committed:
Remove options flow test

marthoc commented:
Assuming tests pass, I’ll just need to update my docs PR to remove reference to hold_temp.

marthoc committed:
Remove hold_temp constants

marthoc commented:
@MartinHjelmare - tests were failing so I tried to rebase against dev and it's pulled in some unrelated commits. I'm going to need to rebase again to drop that commit, sorry if I'm screwing up the history here...

marthoc commented:
@MartinHjelmare Terribly sorry for the mess I made with the botched rebase earlier. I think I've sorted it out and have force pushed a rebase that should make this PR clean. Sorry also to reviewers that had been automatically requested when a bunch of commits got pulled in that shouldn't have.

marthoc committed:
Fix climate tests

marthoc commented:
Not sure why pylint is failing, seems to be a bunch of warnings unrelated to this code.

MartinHjelmare commented:
Looks good! We just need to make the build pass. There seems to be trouble on dev branch at the moment.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+            _LOGGER.debug(
+                "No valid ecobee.conf configuration found for import, delegating to user step"
+            )
+            return await self.async_step_user()
```
We should pass the api key from config yaml here, right?

marthoc replied:
I’m fine with changing that if you want to,, but my thinking was that we should always show the first step (with the API key filled in on the form that is shown on the user step as is done currently) just in case the user wants to change it without restarting Home Assistant (as they would if the key was hard coded in config.yaml and we skipped showing the form by passing the key in as user_input to async_step_user).

MartinHjelmare replied:
If the user uses config yaml they don't want to use the GUI. If they want to use the GUI, the user can remove the config yaml section.

marthoc replied:
Unfortunately they're still going to need to use the GUI since we need to present them the token to authorize on ecobee.com - as they had to do previously via the `configurator` component. I'll make the changes to pass the api key to the user step.

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,112 @@
+                    CONF_REFRESH_TOKEN: ecobee.refresh_token,
+                },
+            )
+        return await self.async_step_user()
```
As above.

marthoc committed:
Pass api key to user step in import flow

FuzzyMistborn commented:
@marthoc just FYI, with my update to https://github.com/nkgilley/python-ecobee-api the release was 0.1.3, not 0.1.2 like you have in the requirements/manifest.json.

marthoc committed:
Update test_config_flow.py

Ensure that the import step calls the user step with the user's api key as user input if importing from ecobee.conf/validating imported keys fails.

marthoc commented:
> @marthoc just FYI, with my update to https://github.com/nkgilley/python-ecobee-api the release was 0.1.3, not 0.1.2 like you have in the requirements/manifest.json.

@FuzzyMistborn that's fine - the changes I needed for this PR are in 0.1.2. After this is merged I will be opening another PR to add create_vacation and delete_vacation services and in that PR I will bump python-ecobee-api to 0.1.3.

FuzzyMistborn commented:
Just saying i don't think there was a 0.1.2 release which is why I point it out.

marthoc commented:
> Just saying i don't think there was a 0.1.2 release which is why I point it out, so PyPi won't be able to find 0.1.2.

There was: https://pypi.org/project/python-ecobee-api/#history  👍

MartinHjelmare commented on homeassistant/components/ecobee/config_flow.py:
```diff
@@ -0,0 +1,120 @@
+            )
+            return await self.async_step_user(
+                user_input={
+                    CONF_API_KEY: self.hass.data[DATA_ECOBEE_CONFIG].get(CONF_API_KEY)
```
What happens if the api key is `None`? Should we check for that before passing the key to the user step?

marthoc replied:
If the key is none (i.e. hasn't been supplied in configuration.yaml), the user step will attempt to validate it, fail, and the user step form will be shown to the user to start the flow fresh. For that reason I don't think it's necessary to validate that the api_key is not none in the import flow step.

marthoc replied:
The same is true if the user supplies an invalid key in configuration.yaml - user step will attempt to validate and fail and the form will be shown. As I see it None is just a subspecies of invalid key that we are already equipped to handle.

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Can be merged when build passes.

marthoc commented:
Looks like we are good to go!

FuzzyMistborn commented:
I tried testing this in my dev instance (fresh install, no ecobee in config.yaml) and every time I click the box to add the integration, i see a new window but it closes and i'm back at the integration menu.  Tried in Chrome and Firefox.  Adding it to config.yaml it worked fine.

First time testing something in dev so wasn't sure if I should mention it here or open a new issue.

frenck commented:
@FuzzyMistborn, in general, opening an issue is a good idea. This allows us to track things easier.
